doc/run-partial.md

```diff
@@ -1,3 +1,82 @@
+
+```js
+function runPartialRecursive(context, options = {}, win = window) {
+  const { axe, document } = win;
+  // Find all frames in context, and determine what context object to use in that frame
+  const frameContexts = axe.utils.getFrameContexts(context);
+  // Run the current context, in the current window.
+  const promiseResults = [ axe.runPartial(context, options) ];
+
+  // Loop over all frames in context
+  frameContexts.forEach(({ frameSelector, frameContext }) =>  {
+    // Find the window of the frame
+    const frame = axe.utils.shadowSelect(frameSelector);
+    const frameWin = frame.contentWindow;
+    // Call axe.runPartial() in the frame, and its child frame
+    const frameResults = runPartialRecursive(frameContext, options, frameWin);
+    promiseResults.push(...frameResults);
+  });
+  return promiseResults;
+};
+```
+
+**important**: The order in which these methods are called matters for performance. Internally, axe-core constructs a flattened tree when `axe.utils.getFrameContexts` is called. This is fairly slow, and so should not happen more than once per frame. When `axe.runPartial` is called, that tree will be used if it still exists. Since this tree can get out of sync with the actual DOM, it is important to call `axe.runPartial` immediately after `axe.utils.getFrameContexts`.
+
+To run efficiently, `axe.runPartial` calls should happen in parallel, so that when possible browsers can test multiple frames simultaneously.
+
+## axe.finishRun(partialResults, options): Promise<AxeResults>
+
+The `axe.finishRun` method does two things: It calls the `after` methods of checks that have them, and it creates the final report. For `axe.finishRun` to work, the order in which the partial results are passed to it is important. Starting from the top window, partial results must include child frame content before sibling frames. For example:
+
+```js
+// Given the following frame structure:
+// top
+// - frame_1
+//   - frame_1a
+// - frame_2
+// The partial results are passed in the following order:
+axe.finishRun([ top, frame_1, frame_1a, frame_2 ])
+```
+
+If for some reason `axe.runPartial` fails to run, the integration API **must** include `null` in the data in place of the results object, so that axe-core knows to skip over it. If a frame fails to run, results from any descending frames **must be omitted**. To illustrate this, consider the following:
+
+```js
+// Given the following frame structure:
+// top
+// - frame_1
+//   - frame_1a
+// - frame_2
+
+// If axe.runPartial throws an error, the results must be passed to finishRun like this:
+axe.finishRun([
+  top, null, /* nothing for frame 1a, */ frame_2
+])
+```
+
+**important**: Since `axe.finishRun` may have access to cross-origin information, it should only be called in an environment that is known not to have third-party scripts. When using a browser driver, this can for example by done in a blank page.
+
```

Comment:
```suggestion

## axe.utils.getFrameContexts(context): FrameContext[]

The `axe.utils.getFrameContexts` method takes any valid context, and returns an array of objects. Each object represents a frame that is in the context. The object has the following properties:

- `frameSelector`: This is a CSS selector, or array of CSS selectors in case of nodes in a shadow DOM tree to locate the frame element to be tested.
- `frameContext`: This is an object is a context object that should be tested in the particular frame.
```